### ISSUE TITLE
tf.data.Dataset.cache('filename') loads entire dataset into memory

### ISSUE THREAD

**Issue by grofte** [tensorflow/tensorflow#50579]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- TensorFlow installed from (source or binary): bin (both pip and docker)
- TensorFlow version (use command below): happens on 2.4.1, 2.5.0 and nightly (2.7.0-dev20210630)
- Python version: 3.8.5

**Describe the current behavior**
When caching a data set on local disk Tensorflow first loads the entire data set into memory. This obviously crashes the job since we are using Dataset and caching on disk because of memory restrictions.

**Describe the expected behavior**
I expected Tensorflow to continuously write to disk while reading the data set from where ever it is coming from. In a streaming manner.  
Alternatively, I could imagine a chunking approach where Tensorflow writes multiple files in the cache to limit the amount of data held in memory at any one time (or writes multiple times to the same file).

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no): no

**Standalone code to reproduce the issue**
[gist](https://gist.github.com/grofte/f617f12a8abef1d085ecaf64500be8b4)

```python
import tensorflow as tf
import tensorflow_datasets as tfds

# This data set is too big to run in Colaboratory, ~30 GB
train = tfds.load(name="bair_robot_pushing_small:2.0.0", split="train",
                  shuffle_files=False, data_dir='./tensorflow_datasets')
train = train.cache('./cache/train')

# We need to do some kind of operation over the data to trigger the caching
batch_size = 64
n = train.batch(batch_size).reduce(0, lambda accum, _: accum + batch_size)
print(n)
```

This is of course a minimal example. We first noticed this problem when loading our from TFRecords files. Caching on file is bad for the training data since you lose a lot of shuffling ability but it's fine for the validation data. If we cache it on a local disk then we can avoid network traffic if the data is elsewhere and the local disk is faster too in our case. So it is still useful - just not when it crashes your job. And really, we could run these training jobs faster if we cached the validation data in memory (but several jobs run on the same machines so we just have to get lucky and not have two RAM intensive parts running at the same time).

**Comment by grofte** [comment]
**Tiny feature request:**  
It would be cool if we could shuffle from cached data. Currently there's two ways to shuffle tf.data.Dataset: 1) shuffle the file paths 2) build a shuffle buffer. 1) is a problem because then I can't apply transformations and cache them (or exploit moving data) without doing all of the serialization and saving and loading and 2) is a problem because we are already using tf.data.Dataset because we run into memory limitations (and honestly a shuffle buffer is not a great shuffle). I am aware that problem 1) simply requires me to write and maintain more code but I thought the point of tf.data was to have clean data streams and transformations. And I think it works pretty well once you get used to it. Problem 2) is also weird because if you cache in memory then I still think you have to use a shuffle buffer - so now your data is two places in memory?

**Comment by rmothukuru** [comment]
@grofte,
We couldn't reproduce the issue as the **`Dataset`** is very huge (30 GB) but your point is taken.

Regarding **`Feature Request`**, we request you to file it as a separate issue so that it can be tracked separately. Thanks!

**Comment by grofte** [comment]
I chose a large data set for exactly that reason. If the data set fits in memory then it will be cached on file fine. But there's no incentive to cache a data set on file if it fits in memory. I will have a look at filing a feature request and see if I can link it here. I think both can be closed in one go though.

**Comment by aaudiber** [comment]
Hi @grofte, the [snapshot](https://www.tensorflow.org/api_docs/python/tf/data/experimental/snapshot) transformation could be closer to what you're looking for. Instead of storing all data in a single file like `cache`, it shards the data into a configurable number of files, and you can shuffle the files in a different way each time you read from the snapshot.

**Comment by grofte** [comment]
@aaudiber Perfect Andrew! That is exactly what I was asking for. 

And I see that it is even out of experimental in the nightly build https://www.tensorflow.org/api_docs/python/tf/data/Dataset?version=nightly#snapshot
